File: awr/rust/src/didkey.rs
//! `did:key` derivation and parsing for Ed25519 (SPEC §5).

use crate::encoding::{b58_encode, b64_decode, b64url_encode_nopad, multibase_decode};
use crate::json::Value;

/// Multicodec identifier for `ed25519-pub`, unsigned-varint (SPEC §5.1).
pub const ED25519_PUB_MULTICODEC: [u8; 2] = [0xed, 0x01];

pub const DID_KEY_PREFIX: &str = "did:key:";

/// A reason code plus detail, ready to be turned into a §11.1 entry.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct KeyError {
    pub code: &'static str,
    pub detail: String,
}

fn err(code: &'static str, detail: impl Into<String>) -> KeyError {
    KeyError { code, detail: detail.into() }
}

/// Known non-Ed25519 multicodecs.
///
/// IMPLEMENTATION CHOICE (§11.2 registers both `AWR-KEY-002` "bad multibase,
/// multicodec, or key length" and `AWR-KEY-004` "unsupported key type" without
/// separating them): a DID that decodes cleanly to a *recognised* other key type
/// is `AWR-KEY-004`; anything else structural is `AWR-KEY-002`.
fn named_multicodec(prefix: &[u8]) -> Option<&'static str> {
    let two = |a: u8, b: u8| prefix.len() >= 2 && prefix[0] == a && prefix[1] == b;
    if two(0xec, 0x01) {
        Some("x25519-pub")
    } else if two(0xe7, 0x01) {
        Some("secp256k1-pub")
    } else if two(0x80, 0x24) {
        Some("p256-pub")
    } else if two(0x81, 0x24) {
        Some("p384-pub")
    } else if two(0x82, 0x24) {
        Some("p521-pub")
    } else if two(0xeb, 0x01) {
        Some("bls12_381-g2-pub")
    } else if two(0x85, 0x24) {
        Some("rsa-pub")
    } else {
        None
    }
}

/// The multibase string that follows `did:key:` — also the fragment used in
/// `verificationMethod` (§5.3).
pub fn method_specific_id(public_key: &[u8; 32]) -> String {
    let mut multicodec = Vec::with_capacity(34);
    multicodec.extend_from_slice(&ED25519_PUB_MULTICODEC);
    multicodec.extend_from_slice(public_key);
    let mut s = String::with_capacity(48);
    s.push('z');
    s.push_str(&b58_encode(&multicodec));
    s
}

/// `did:key:z…` for an Ed25519 public key (§5.1).
pub fn did_from_public_key(public_key: &[u8; 32]) -> String {
    format!("{}{}", DID_KEY_PREFIX, method_specific_id(public_key))
}

/// `did:key:z…#z…` (§5.3).
pub fn verification_method_for(public_key: &[u8; 32]) -> String {
    let msi = method_specific_id(public_key);
    format!("{}{}#{}", DID_KEY_PREFIX, msi, msi)
}

/// Decode an Ed25519 public key from a `did:key` DID.
///
/// * a value that is not a `did:key` at all → `AWR-KEY-001`
/// * bad multibase / multicodec / length → `AWR-KEY-002`
/// * a well-formed `did:key` for a different key type → `AWR-KEY-004`
pub fn parse_did_key(did: &str) -> Result<[u8; 32], KeyError> {
    if !did.starts_with(DID_KEY_PREFIX) {
        return Err(err(
            "AWR-KEY-001",
            format!(
                "issuer identifier {:?} is not a did:key; AWR/2 supports no other DID method or an HTTPS issuer (§5.1)",
                did
            ),
        ));
    }
    let msi = &did[DID_KEY_PREFIX.len()..];
    if msi.is_empty() {
        return Err(err("AWR-KEY-002", "did:key with an empty method-specific identifier"));
    }
    // IMPLEMENTATION CHOICE (§5.1 gives the did:key form without saying whether
    // a path, query or fragment may follow): they are rejected as
    // `AWR-KEY-002`, because §5.3 builds verificationMethod by appending
    // `#<msi>` to issuer.id, and a fragment already present there would make the
    // two statements of the key disagree silently.
    if let Some(bad) = msi.chars().find(|c| matches!(c, '#' | '?' | '/')) {
        return Err(err(
            "AWR-KEY-002",
            format!("issuer.id must be a bare did:key; it contains `{}`", bad),
        ));
    }
    let bytes = multibase_decode(msi)
        .map_err(|e| err("AWR-KEY-002", format!("did:key multibase decode failed: {}", e)))?;
    if bytes.len() < 2 {
        return Err(err("AWR-KEY-002", "did:key decodes to fewer than 2 bytes"));
    }
    if bytes[0..2] != ED25519_PUB_MULTICODEC {
        if let Some(name) = named_multicodec(&bytes) {
            return Err(err(
                "AWR-KEY-004",
                format!(
                    "did:key names a {} key; AWR/2 defines Ed25519 (0xed 0x01) only (§5.1)",
                    name
                ),
            ));
        }
        return Err(err(
            "AWR-KEY-002",
            format!(
                "multicodec prefix 0x{:02x} 0x{:02x} is not ed25519-pub (0xed 0x01)",
                bytes[0], bytes[1]
            ),
        ));
    }
    let key = &bytes[2..];
    if key.len() != 32 {
        return Err(err(
            "AWR-KEY-002",
            format!("Ed25519 public key must be 32 bytes, got {}", key.len()),
        ));
    }
    let mut out = [0u8; 32];
    out.copy_from_slice(key);
    Ok(out)
}

/// Check an optional `issuer.publicKeyJwk` against the DID-derived key (§5.2).
///
/// A mismatch is `AWR-KEY-003` and invalidates the document: two disagreeing
/// statements of the signing key inside one signed document are a downgrade
/// surface, not a redundancy.
pub fn check_public_key_jwk(jwk: &Value, expected: &[u8; 32]) -> Result<(), KeyError> {
    let obj = match jwk {
        Value::Object(_) => jwk,
        _ => {
            return Err(err(
                "AWR-KEY-003",
                format!("issuer.publicKeyJwk is a {}, not an object", jwk.type_name()),
            ))
        }
    };
    match obj.get("kty").and_then(|v| v.as_str()) {
        Some("OKP") => {}
        Some(other) => {
            return Err(err(
                "AWR-KEY-003",
                format!("publicKeyJwk.kty is {:?}, expected \"OKP\" (RFC 8037)", other),
            ))
        }
        None => return Err(err("AWR-KEY-003", "publicKeyJwk.kty missing")),
    }
    match obj.get("crv").and_then(|v| v.as_str()) {
        Some("Ed25519") => {}
        Some(other) => {
            return Err(err(
                "AWR-KEY-003",
                format!("publicKeyJwk.crv is {:?}, expected \"Ed25519\"", other),
            ))
        }
        None => return Err(err("AWR-KEY-003", "publicKeyJwk.crv missing")),
    }
    let x = match obj.get("x").and_then(|v| v.as_str()) {
        Some(x) => x,
        None => return Err(err("AWR-KEY-003", "publicKeyJwk.x missing or not a string")),
    };
    let raw = b64_decode(x, false)
        .map_err(|e| err("AWR-KEY-003", format!("publicKeyJwk.x is not base64url: {}", e)))?;
    if raw.as_slice() != expected.as_slice() {
        return Err(err(
            "AWR-KEY-003",
            format!(
                "publicKeyJwk.x decodes to a different key than issuer.id: {} vs {}",
                x,
                b64url_encode_nopad(expected)
            ),
        ));
    }
    Ok(())
}

/// Build an RFC 8037 OKP/Ed25519 JWK for a public key (used by `issue`).
pub fn public_key_jwk(public_key: &[u8; 32]) -> Value {
    Value::object(vec![
        ("kty".to_string(), Value::string("OKP")),
        ("crv".to_string(), Value::string("Ed25519")),
        ("x".to_string(), Value::string(b64url_encode_nopad(public_key))),
    ])
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::json::parse;

    #[test]
    fn derivation_shape_matches_spec() {
        // §5.1: the method-specific identifier is 48 characters and begins z6Mk.
        // Checked over many keys, since the property is a consequence of the
        // 0xed 0x01 prefix rather than of any single key.
        for seed in 0u8..32 {
            let pk = [seed; 32];
            let msi = method_specific_id(&pk);
            assert_eq!(msi.len(), 48, "key {:?} produced {}", seed, msi);
            assert!(msi.starts_with("z6Mk"), "key {:?} produced {}", seed, msi);
            let did = did_from_public_key(&pk);
            assert_eq!(parse_did_key(&did).unwrap(), pk);
            assert_eq!(verification_method_for(&pk), format!("{}#{}", did, msi));
        }
    }

    #[test]
    fn rejects_non_didkey_and_malformed() {
        assert_eq!(parse_did_key("https://example.com/issuer").unwrap_err().code, "AWR-KEY-001");
        assert_eq!(parse_did_key("did:web:example.com").unwrap_err().code, "AWR-KEY-001");
        assert_eq!(parse_did_key("did:key:").unwrap_err().code, "AWR-KEY-002");
        // base64 multibase prefix instead of base58btc
        assert_eq!(parse_did_key("did:key:mSGVsbG8").unwrap_err().code, "AWR-KEY-002");
        // wrong length: 31-byte key
        let short = {
            let mut v = ED25519_PUB_MULTICODEC.to_vec();
            v.extend_from_slice(&[7u8; 31]);
            format!("did:key:z{}", crate::encoding::b58_encode(&v))
        };
        assert_eq!(parse_did_key(&short).unwrap_err().code, "AWR-KEY-002");
        // fragment inside issuer.id
        let pk = [3u8; 32];
        let with_frag = format!("{}#{}", did_from_public_key(&pk), method_specific_id(&pk));
        assert_eq!(parse_did_key(&with_frag).unwrap_err().code, "AWR-KEY-002");
    }

    #[test]
    fn rejects_other_key_types_as_key_004() {
        let mut v = vec![0x80, 0x24];
        v.extend_from_slice(&[1u8; 33]);
        let did = format!("did:key:z{}", crate::encoding::b58_encode(&v));
        let e = parse_did_key(&did).unwrap_err();
        assert_eq!(e.code, "AWR-KEY-004");
        assert!(e.detail.contains("p256-pub"));
    }

    #[test]
    fn jwk_consistency() {
        let pk = [9u8; 32];
        assert!(check_public_key_jwk(&public_key_jwk(&pk), &pk).is_ok());
        let other = [8u8; 32];
        assert_eq!(check_public_key_jwk(&public_key_jwk(&other), &pk).unwrap_err().code, "AWR-KEY-003");
        let bad_kty = parse(br#"{"kty":"EC","crv":"Ed25519","x":"AAAA"}"#).unwrap();
        assert_eq!(check_public_key_jwk(&bad_kty, &pk).unwrap_err().code, "AWR-KEY-003");
        let bad_crv = parse(br#"{"kty":"OKP","crv":"X25519","x":"AAAA"}"#).unwrap();
        assert_eq!(check_public_key_jwk(&bad_crv, &pk).unwrap_err().code, "AWR-KEY-003");
        let no_x = parse(br#"{"kty":"OKP","crv":"Ed25519"}"#).unwrap();
        assert_eq!(check_public_key_jwk(&no_x, &pk).unwrap_err().code, "AWR-KEY-003");
    }
}

File: awr/reference/python/awr/didkey.py
"""``did:key`` for Ed25519, and the signing-key handle (SPEC.md section 5).

``did:key:z<base58btc(0xed 0x01 || publicKey)>``.  ``0xed 0x01`` is the unsigned-varint
multicodec identifier for ``ed25519-pub``; the ``z`` is multibase base58btc.  A verifier
derives the key from the DID, so no network lookup and no key selection ever happens.
"""

from __future__ import annotations

import base64
import binascii
import json
import re
from typing import Any, Dict, Optional

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    NoEncryption,
    PrivateFormat,
    PublicFormat,
)

from .errors import (
    AWR_KEY_001,
    AWR_KEY_002,
    AWR_KEY_003,
    AWR_KEY_004,
    AwrError,
)
from .multibase import (
    b58decode,
    multibase_decode_base58btc,
    multibase_encode_base58btc,
)

DID_KEY_PREFIX = "did:key:"

#: Unsigned-varint multicodec for ed25519-pub.
MULTICODEC_ED25519_PUB = b"\xed\x01"
#: Unsigned-varint multicodec for ed25519-priv, used by the optional key-file form.
MULTICODEC_ED25519_PRIV = b"\x80\x26"

#: Multicodecs that name a real public key of a type AWR/2 does not support.  Telling
#: these apart from garbage is the difference between AWR-KEY-004 and AWR-KEY-002.
KNOWN_OTHER_KEY_MULTICODECS = {
    b"\xe7\x01": "secp256k1-pub",
    b"\xec\x01": "x25519-pub",
    b"\xea\x01": "bls12_381-g1-pub",
    b"\xeb\x01": "bls12_381-g2-pub",
    b"\x80\x24": "p256-pub",
    b"\x81\x24": "p384-pub",
    b"\x82\x24": "p521-pub",
    b"\x85\x24": "rsa-pub",
}

_DID_KEY_MSI_RE = re.compile(r"^z[1-9A-HJ-NP-Za-km-z]+$")


def derive_did_key(public_key_bytes: bytes) -> str:
    """Derive the ``did:key`` DID of a raw 32-byte Ed25519 public key."""
    if len(public_key_bytes) != 32:
        raise ValueError(
            "an Ed25519 public key is 32 bytes, got %d" % (len(public_key_bytes),)
        )
    return DID_KEY_PREFIX + multibase_encode_base58btc(
        MULTICODEC_ED25519_PUB + bytes(public_key_bytes)
    )


def method_specific_id(did: str) -> str:
    """The part of a ``did:key`` after the method prefix (the ``z...`` multibase string)."""
    if not isinstance(did, str) or not did.startswith(DID_KEY_PREFIX):
        raise AwrError(AWR_KEY_001, "issuer.id %r is not a did:key" % (did,))
    return did[len(DID_KEY_PREFIX):]


def parse_did_key(did: Any) -> bytes:
    """Return the raw 32-byte Ed25519 public key named by *did*.

    Raises ``AWR-KEY-001`` when this is not a ``did:key`` at all, ``AWR-KEY-004`` when it
    is a well-formed ``did:key`` for a key type AWR/2 does not support, and
    ``AWR-KEY-002`` on any other deviation (section 5.1).
    """
    if not isinstance(did, str):
        raise AwrError(
            AWR_KEY_001, "issuer.id must be a string, got %s" % type(did).__name__
        )
    if not did.startswith(DID_KEY_PREFIX):
        raise AwrError(AWR_KEY_001, "issuer.id %r is not a did:key" % (did,))
    msi = did[len(DID_KEY_PREFIX):]
    if "#" in msi or "?" in msi or "/" in msi:
        raise AwrError(
            AWR_KEY_002,
            "did:key %r carries a fragment, path or query; issuer.id must be the bare DID"
            % (did,),
        )
    if not msi.startswith("z"):
        raise AwrError(
            AWR_KEY_002,
            "did:key multibase prefix must be 'z' (base58btc), got %r" % (msi[:1],),
        )
    if _DID_KEY_MSI_RE.match(msi) is None:
        raise AwrError(
            AWR_KEY_002,
            "did:key method-specific id %r is not base58btc" % (msi,),
        )
    try:
        decoded = multibase_decode_base58btc(msi)
    except ValueError as exc:
        raise AwrError(AWR_KEY_002, "did:key is not decodable: %s" % (exc,))
    if len(decoded) < 2:
        raise AwrError(AWR_KEY_002, "did:key payload is shorter than its multicodec")
    multicodec = decoded[:2]
    key_bytes = decoded[2:]
    if multicodec == MULTICODEC_ED25519_PUB:
        if len(key_bytes) != 32:
            raise AwrError(
                AWR_KEY_002,
                "ed25519-pub multicodec with %d key bytes, expected 32"
                % (len(key_bytes),),
            )
        return key_bytes
    other = KNOWN_OTHER_KEY_MULTICODECS.get(multicodec)
    if other is not None:
        raise AwrError(
            AWR_KEY_004,
            "did:key names a %s key; AWR/2 supports only ed25519-pub (0xed 0x01)"
            % (other,),
        )
    raise AwrError(
        AWR_KEY_002,
        "unrecognised multicodec 0x%s in did:key" % (multicodec.hex(),),
    )


def verification_method_for(did: str) -> str:
    """``<did>#<method-specific-id>`` (section 5.3)."""
    return "%s#%s" % (did, method_specific_id(did))


def b64url_decode(text: str) -> bytes:
    """Decode unpadded or padded base64url."""
    if not isinstance(text, str):
        raise ValueError("base64url value must be a string")
    padding = "=" * (-len(text) % 4)
    try:
        return base64.urlsafe_b64decode(text + padding)
    except (binascii.Error, ValueError) as exc:
        raise ValueError("not valid base64url: %s" % (exc,))


def b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def check_public_key_jwk(jwk: Any, public_key_bytes: bytes) -> None:
    """Check an optional ``issuer.publicKeyJwk`` against the ``did:key`` (section 5.2).

    Any disagreement is ``AWR-KEY-003`` and invalidates the document: two statements of
    the signing key inside one signed document is a downgrade surface, not redundancy.
    """
    if not isinstance(jwk, dict):
        raise AwrError(
            AWR_KEY_003,
            "publicKeyJwk must be an object, got %s" % type(jwk).__name__,
        )
    if jwk.get("kty") != "OKP":
        raise AwrError(
            AWR_KEY_003, "publicKeyJwk.kty must be 'OKP', got %r" % (jwk.get("kty"),)
        )
    if jwk.get("crv") != "Ed25519":
        raise AwrError(
            AWR_KEY_003,
            "publicKeyJwk.crv must be 'Ed25519', got %r" % (jwk.get("crv"),),
        )
    x = jwk.get("x")
    if not isinstance(x, str):
        raise AwrError(AWR_KEY_003, "publicKeyJwk.x missing or not a string")
    try:
        decoded = b64url_decode(x)
    except ValueError as exc:
        raise AwrError(AWR_KEY_003, "publicKeyJwk.x does not decode: %s" % (exc,))
    if decoded != public_key_bytes:
        raise AwrError(
            AWR_KEY_003,
            "publicKeyJwk.x names a different key than issuer.id",
        )


def public_key_from_bytes(public_key_bytes: bytes) -> Ed25519PublicKey:
    return Ed25519PublicKey.from_public_bytes(public_key_bytes)


def verify_signature(public_key_bytes: bytes, signature: bytes, message: bytes) -> bool:
    """Pure Ed25519 verification (RFC 8032)."""
    try:
        public_key_from_bytes(public_key_bytes).verify(signature, message)
    except (InvalidSignature, ValueError):
        return False
    return True


class SigningKey:
    """An Ed25519 private key with its AWR identity.

    ``SigningKey`` is the only way to sign in this package, and it is deliberately unable
    to produce an AWR/1 proof (section 12: an implementation MUST NOT issue AWR/1).
    """

    def __init__(self, private_key: Ed25519PrivateKey) -> None:
        self._private = private_key
        self._public_bytes = private_key.public_key().public_bytes(
            Encoding.Raw, PublicFormat.Raw
        )

    # -- construction ---------------------------------------------------------

    @classmethod
    def generate(cls) -> "SigningKey":
        return cls(Ed25519PrivateKey.generate())

    @classmethod
    def from_seed(cls, seed: bytes) -> "SigningKey":
        if len(seed) != 32:
            raise ValueError("an Ed25519 seed is 32 bytes, got %d" % (len(seed),))
        return cls(Ed25519PrivateKey.from_private_bytes(bytes(seed)))

    @classmethod
    def from_jwk(cls, jwk: Dict[str, Any]) -> "SigningKey":
        if not isinstance(jwk, dict):
            raise ValueError("JWK must be an object")
        if jwk.get("kty") != "OKP" or jwk.get("crv") != "Ed25519":
            raise ValueError("key file must be an RFC 8037 OKP/Ed25519 JWK")
        d = jwk.get("d")
        if not isinstance(d, str):
            raise ValueError("JWK is a public key: no 'd' member")
        key = cls.from_seed(b64url_decode(d))
        x = jwk.get("x")
        if isinstance(x, str) and b64url_decode(x) != key.public_key_bytes:
            raise ValueError("JWK 'x' does not match the key derived from 'd'")
        return key

    @classmethod
    def from_key_file_text(cls, text: str) -> "SigningKey":
        """Load a private key from the CLI ``--key`` file.

        Accepted forms (section 17 does not specify one; see README):

        * an RFC 8037 OKP/Ed25519 JWK object with ``d``
        * ``{"privateKeySeedHex": "<64 hex chars>"}``
        * ``{"privateKeyMultibase": "z<base58btc(0x80 0x26 || seed)>"}``
        * a bare 64-hex-character seed on a single line
        """
        stripped = text.strip()
        if stripped.startswith("{"):
            try:
                data = json.loads(stripped)
            except ValueError as exc:
                raise ValueError("key file is not valid JSON: %s" % (exc,))
            if not isinstance(data, dict):
                raise ValueError("key file JSON must be an object")
            if "kty" in data:
                return cls.from_jwk(data)
            for member in ("privateKeySeedHex", "seedHex"):
                if member in data:
                    return cls.from_seed(bytes.fromhex(str(data[member])))
            if "privateKeyMultibase" in data:
                decoded = multibase_decode_base58btc(str(data["privateKeyMultibase"]))
                if decoded[:2] != MULTICODEC_ED25519_PRIV:
                    raise ValueError(
                        "privateKeyMultibase multicodec is not ed25519-priv (0x80 0x26)"
                    )
                return cls.from_seed(decoded[2:])
            raise ValueError(
                "key file object has none of kty / privateKeySeedHex / "
                "privateKeyMultibase"
            )
        if stripped.startswith("z"):
            decoded = b58decode(stripped[1:])
            if decoded[:2] == MULTICODEC_ED25519_PRIV:
                return cls.from_seed(decoded[2:])
            raise ValueError("multibase key file is not an ed25519-priv value")
        try:
            raw = bytes.fromhex(stripped)
        except ValueError:
            raise ValueError(
                "key file is neither JSON, a multibase value, nor a hex seed"
            )
        return cls.from_seed(raw)

    # -- identity -------------------------------------------------------------

    @property
    def public_key_bytes(self) -> bytes:
        return self._public_bytes

    @property
    def did(self) -> str:
        return derive_did_key(self._public_bytes)

    @property
    def verification_method(self) -> str:
        return verification_method_for(self.did)

    def public_key_jwk(self) -> Dict[str, str]:
        return {
            "kty": "OKP",
            "crv": "Ed25519",
            "x": b64url_encode(self._public_bytes),
        }

    def private_key_jwk(self) -> Dict[str, str]:
        seed = self._private.private_bytes(
            Encoding.Raw, PrivateFormat.Raw, NoEncryption()
        )
        jwk = self.public_key_jwk()
        jwk["d"] = b64url_encode(seed)
        return jwk

    def seed_hex(self) -> str:
        return self._private.private_bytes(
            Encoding.Raw, PrivateFormat.Raw, NoEncryption()
        ).hex()

    # -- signing --------------------------------------------------------------

    def sign(self, message: bytes) -> bytes:
        """Pure Ed25519 signature over *message* (RFC 8032), 64 bytes."""
        return self._private.sign(message)


def load_key_file(path: str) -> SigningKey:
    with open(path, "r", encoding="utf-8") as handle:
        return SigningKey.from_key_file_text(handle.read())


def optional_public_key_from_jwk(jwk: Any) -> Optional[bytes]:
    """Best-effort raw key from a JWK, used by the AWR/1 legacy path (section 12)."""
    if not isinstance(jwk, dict):
        return None
    x = jwk.get("x")
    if not isinstance(x, str):
        return None
    try:
        decoded = b64url_decode(x)
    except ValueError:
        return None
    return decoded if len(decoded) == 32 else None
